### PR TITLE
feat: Implement Support for Validator Network

### PR DESCRIPTION
**Pull request by MatusKysel**
>CI
{
  "hardfork": "maxwell",
  "bsc": "https://github.com/bnb-chain/bsc/commit/d173feee3d55a3f015aa5aa46c9d271cbd56496d"
}
>CI

## Summary
This PR enhances the StakeHub contract by implementing NodeID registration functionality and extending authorization to support BEP-410 agents. These changes enable validators to manage their network identities and allow authorized agents to perform NodeID operations on behalf of validators.

### NodeID Management
- Added state variables to track validator NodeIDs:
  - `maxNodeIDs`: Governance-controlled parameter to limit the number of NodeIDs per validator (default: 5)
  - `validatorNodeIDs`: Mapping from validator address to an array of NodeIDs

- Implemented key functions for NodeID management:
  - `addNodeID`: Allows validators to register new NodeIDs with duplicate checking
  - `removeNodeID`: Allows validators to remove NodeIDs with efficient swap-and-pop technique
  - `replaceNodeID`: New function to efficiently replace existing NodeIDs
  - `getNodeIDs`: Returns all NodeIDs for a given validator
  - `listNodeIDsFor`: Batch query to return NodeIDs for multiple validators

- Added storage optimization:
  - Automatic cleanup when a validator's NodeID array becomes empty

- Added appropriate events for tracking changes:
  - `NodeIDAdded`: Emitted when a new NodeID is registered
  - `NodeIDRemoved`: Emitted when a NodeID is removed

### Authorization Enhancement
- Extended the `onlyAuthorized` modifier to support BEP-410 agents:
  - Now accepts three types of authorized callers:
    1. The validator itself
    2. The validator's consensus address
    3. The validator's BEP-410 authorized agent
  - Maintains security while providing flexibility for validator operations

- Added governance parameter for `maxNodeIDs` to allow adjusting the maximum number of NodeIDs per validator

## Security Considerations
- Comprehensive authorization checks through the enhanced `onlyAuthorized` modifier
- NodeID operations are protected by standard contract modifiers:
  - `whenNotPaused`
  - `notInBlackList`
  - `validatorExist`
- Input validation to prevent invalid operations:
  - Zero NodeID rejection
  - Maximum NodeID limit enforcement
  - Duplicate NodeID prevention
  - Validation that NodeIDs exist before removal or replacement

## Testing
- Extensive tests have been added to verify:
  - All NodeID management functionality
  - BEP-410 agent authorization
  - Edge cases including:
    - Adding maximum number of NodeIDs
    - Removing the last NodeID
    - Replacing existing NodeIDs
    - Unauthorized access attempts
    - Operations on non-existent NodeIDs

This enhancement enables validators to register their network identities and allows authorized agents to manage NodeIDs on their behalf, facilitating secure and efficient validator communication in the BSC network while maintaining the security model of the contract.

### CODE REVIEW

File: contracts/StakeHub.sol [hunk 1]
```diff
@@ -1011,6 +1046,138 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
     }
 
+     /**
+     * @notice Adds a new NodeID (of type bytes32) to the validator's registry.
+     * @param validator The address of the validator.
+     * @param newNodeID The NodeID to be added (must be nonzero).
+     */
+    function addNodeID(
+        address validator,
+        bytes32 newNodeID
```
Comment (galaio) [inline contracts/StakeHub.sol:1056]: Could here support adding multiple nodeIDs?

File: contracts/StakeHub.sol [hunk 2]
```diff
@@ -236,6 +248,24 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
 
+    /**
+     * @dev Modifier to ensure that only an authorized validator can update NodeIDs.
+     * The caller must either be:
+     * 1. The validator itself (i.e. the provided address)
+     * 2. The designated alternative address recorded in `consensusToOperator` for that validator
+     * 3. The BEP-410 authorized sender (agent) for the validator
+     */
+    modifier onlyAuthorized(address validator) {
+        address sender = _bep410MsgSender();
```
Comment (galaio) [inline contracts/StakeHub.sol:259]: The `_bep410MsgSender` will return the agent's operator address or itself. So I think the permission checking should be:

```solidity
require(
            _validatorSet.contains(sender) || 
            consensusToOperator[msg.sender] != address(0),
            "Unauthorized caller"
        );
```

and the param can be removed.

File: contracts/StakeHub.sol [hunk 3]
```diff
@@ -1011,6 +1046,138 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
 
+     /**
+     * @notice Adds a new NodeID (of type bytes32) to the validator's registry.
+     * @param validator The address of the validator.
+     * @param newNodeID The NodeID to be added (must be nonzero).
+     */
+    function addNodeID(
+        address validator,
+        bytes32 newNodeID
+    ) external whenNotPaused notInBlackList validatorExist(validator) onlyAuthorized(validator) {
```
Comment (galaio) [inline contracts/StakeHub.sol:1057]: Should checking `validatorExist` not be needed?

File: contracts/StakeHub.sol [hunk 4]
```diff
@@ -1011,6 +1046,138 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+    }
+
+    /**
+     * @notice Removes a specified NodeID from the validator's registry.
+     * @param validator The address of the validator.
+     * @param targetNodeID The NodeID to be removed.
+     */
+    function removeNodeID(
+        address validator,
+        bytes32 targetNodeID
```
Comment (galaio) [inline contracts/StakeHub.sol:1083]: Same question here. Could here support adding multiple nodeIDs?

File: contracts/StakeHub.sol [hunk 5]
```diff
@@ -1011,6 +1046,138 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+        }
+    }
+
+    /**
+     * @notice Replaces an existing NodeID with a new one for a validator.
+     * @param validator The address of the validator.
+     * @param oldNodeID The NodeID to be replaced.
+     * @param newNodeID The new NodeID to replace the old one with.
+     */
+    function replaceNodeID(
```
Comment (galaio) [inline contracts/StakeHub.sol:1117]: I think the `replaceNodeIDs` logic could be very easy.  input the new adding NodeIDs, and just delete the old NodeIDs, and assigning the new ones.

File: contracts/StakeHub.sol [hunk 6]
```diff
@@ -1011,6 +1046,138 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+        emit NodeIDRemoved(validator, oldNodeID);
+        emit NodeIDAdded(validator, newNodeID);
+    }
+
+    /**
+     * @notice Returns all registered NodeIDs for a given validator.
+     * @param validator The address of the validator.
+     * @return An array of NodeIDs (bytes32[]).
+     */
+    function getNodeIDs(
```
Comment (galaio) [inline contracts/StakeHub.sol:1159]: I think this method can be removed, because of the `listNodeIDsFor` can cover the secenario.

File: contracts/StakeHub.sol [hunk 7]
```diff
@@ -1011,6 +1046,138 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+    ) external view returns (bytes32[] memory) {
+        return validatorNodeIDs[validator];
+    }
+
+    /**
+     * @notice Returns all validators with their registered NodeIDs.
+     * @param validatorsToQuery The addresses of the validators.
+     * @return An array of NodeIDs (bytes32[][]).
+     */
+    function listNodeIDsFor(
```
Comment (galaio) [inline contracts/StakeHub.sol:1170]: By the way, could you help add some UTs to check these logics?

File: contracts/StakeHub.sol (hunk 2 above)
Comment (galaio) [reply contracts/StakeHub.sol:259]: But the checking seems wrong, if the `msg.sender` is an attack address, it also pass the `sender == validator` checking.

Comment (galaio) [reply contracts/StakeHub.sol:265]: I think we can simplify the permission checking~ like this
```solidity
require(
            _validatorSet.contains(sender) || 
            consensusToOperator[msg.sender] != address(0),
            "Unauthorized caller"
        );
```

File: contracts/StakeHub.sol [hunk 8]
```diff
@@ -133,6 +139,14 @@ contract StakeHub is SystemV2, Initializable, Protectable {
     // agent => validator operator address
     mapping(address => address) public agentToOperator;
 
+    // network related values //
+    // governance controlled maximum number of NodeIDs per validator (default is 5).
+    uint256 public maxNodeIDs = 5;
```
Comment (buddh0) [inline contracts/StakeHub.sol:144]: do solidity support set default value this way？

**Review by pythonberg1997 — DISMISSED** [review]

File: contracts/StakeHub.sol [hunk 9]
```diff
@@ -133,6 +139,14 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+
+    // mapping from a validator's address to an array of their registered NodeIDs,
```
Comment (buddh0) [inline contracts/StakeHub.sol:146]: validator's address 
meaning
validator's consensus address？

File: contracts/StakeHub.sol [hunk 10]
```diff
@@ -133,6 +139,14 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+    // where each NodeID is stored as a fixed 32-byte value.
+    mapping(address => bytes32[]) private validatorNodeIDs;
```
Comment (buddh0) [inline contracts/StakeHub.sol:148]: consensusToNodeIDs
follow the other names in this file, such as `consensusToOperator` or `consensusExpiration`

File: contracts/StakeHub.sol (hunk 2 above)
Comment (galaio) [reply contracts/StakeHub.sol:265]: Emm....I got you.
```solidity
require(
            (_validatorSet.contains(sender) && consensusToOperator[validator]==sender) || // checking operator -> validator relationship, if msg.sneder is operator or agent
            (consensusToOperator[validator] != address(0) && msg.sender == validator), // checking if there exist the validator, if msg.sneder is validator
            "Unauthorized caller"
        );
```

File: contracts/StakeHub.sol (hunk 9 above)
Comment (buddh0) [inline contracts/StakeHub.sol:146]: validator's address meaning 'validator's operator address'?
because only operator address can represent one validator, consensus address could be changed

File: contracts/StakeHub.sol [hunk 11]
```diff
@@ -236,6 +254,24 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
     }
 
+    /**
+     * @dev Modifier to ensure that only an authorized validator can update NodeIDs.
+     * The caller must either be:
+     * 1. The validator itself (i.e. the provided address)
+     * 2. The designated alternative address recorded in `consensusToOperator` for that validator
+     * 3. The BEP-410 authorized sender (agent) for the validator
+     */
+    modifier onlyAuthorized(address validator) {
```
Comment (buddh0) [inline contracts/StakeHub.sol:264]: IMO,
this modifier and interfaces `addNodeIDs`, `removeNodeIDs` no need to pass an param `address validator`,
it should be retrieved from sender.

Comment (buddh0) [reply contracts/StakeHub.sol:264]: define a modifier
```
    function _bep563MsgSender() internal view returns (address) {
        if (consensusToOperator[msg.sender] != address(0)) {
            return consensusToOperator[msg.sender];
        }

        return _bep410MsgSender();
    }
```
then use it the way `_bep410MsgSender` used
such as  `validatorExist(_bep563MsgSender())`

File: contracts/StakeHub.sol [hunk 12]
```diff
@@ -256,6 +274,7 @@ contract StakeHub is SystemV2, Initializable, Protectable {
         downtimeJailTime = 2 days;
         felonyJailTime = 30 days;
         maxFelonyBetweenBreatheBlock = 2;
+        maxNodeIDs = 5;
```
Comment (buddh0) [inline contracts/StakeHub.sol:277]: uint256 public constant INIT_MAX_NUMBER_NODE_ID = 5;
    bool maxNodeIDsInitialized

    function maxNodeIDsInitializer() internal{
        if (!maxNodeIDsInitialized)
        	maxNodeIDs = INIT_MAX_NUMBER_NODE_ID
        	maxNodeIDsInitialized = true
        }
    }
    
    then use `maxNodeIDsInitializer` for new added functions

Comment (buddh0) [reply contracts/StakeHub.sol:277]: `initialize` will not be called when maxwell hardfork enabled.

Comment (galaio) [reply contracts/StakeHub.sol]: How about this, add a getter method, and return the default when empty.

```solidity
function getMaxNodeIDs() external view returns (uint256) {
        if (maxNodeIDs == 0) {
            return 5;
        }
        return turnLength;
    }

```

File: contracts/StakeHub.sol [hunk 13]
```diff
@@ -1011,6 +1035,176 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+     * @notice Adds multiple new NodeIDs to the validator's registry.
+     * @param nodeIDs Array of NodeIDs to be added.
+     */
+    function addNodeIDs( bytes32[] calldata nodeIDs) external whenNotPaused notInBlackList validatorExist(_bep563MsgSender()) {
+        if (nodeIDs.length == 0) {
+            revert InvalidNodeID();
+        }
+
+        address operatorAddress = _bep563MsgSender();
+        bytes32[] storage existingNodeIDs = validatorNodeIDs[operatorAddress];
```
[the remark this answers is not in the document]
Comment (galaio) [reply contracts/StakeHub.sol:1054]: Yeah, you are right~

**Review by buddh0 — DISMISSED** [review]

File: test/utils/interface/IStakeHub.sol [hunk 14]
```diff
@@ -185,4 +187,11 @@ interface StakeHub {
 
     function agentToOperator(address) external view returns (address);
     function updateAgent(address newAgent) external;
+
+    // NodeID management functions
+    function addNodeIDs(bytes32[] calldata newNodeIDs) external;
+    function removeNodeIDs(bytes32[] calldata targetNodeIDs) external;
+    function replaceNodeIDs(bytes32[] calldata newNodeIDs) external;
```
Comment (alex-10072) [inline test/utils/interface/IStakeHub.sol:194]: I think we don't need `replaceNodeIDs()`, `addNodeIDs()` and `removeNodeIDs` would be enough. it can simplify the code, although the validator may send two transactions to perform add and remove actions seperately.

File: contracts/StakeHub.sol [hunk 15]
```diff
@@ -1011,6 +1039,119 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+    function addNodeIDs(
+        bytes32[] calldata nodeIDs
+    ) external whenNotPaused notInBlackList validatorExist(_bep563MsgSender()) {
+        maxNodeIDsInitializer();
+
+        if (nodeIDs.length == 0) {
+            revert InvalidNodeID();
+        }
+
+        address operatorAddress = _bep563MsgSender();
```
Comment (zzzckck) [inline contracts/StakeHub.sol:1055]: pls note: one operator address can maintain several validators

File: contracts/StakeHub.sol [hunk 16]
```diff
@@ -1011,6 +1039,119 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+        bytes32[] storage existingNodeIDs = validatorNodeIDs[operatorAddress];
+        uint256 currentLength = existingNodeIDs.length;
+
+        if (currentLength + nodeIDs.length > maxNodeIDs) {
```
Comment (zzzckck) [inline contracts/StakeHub.sol:1059]: pls note: one operator address can maintain several validators

File: contracts/StakeHub.sol (hunk 15 above)
Comment (buddh0) [reply contracts/StakeHub.sol:1055]: > pls note: one operator address can maintain several validators

really?
```
   // validator operator address => validator info
    mapping(address => Validator) private _validators;
```

File: contracts/StakeHub.sol [hunk 17]
```diff
@@ -84,6 +86,12 @@ contract StakeHub is SystemV2, Initializable, Protectable {
     error InvalidAgent();
     // @notice signature: 0x682a6e7c
     error InvalidValidator();
+    // @notice signature: 0x12345678
+    error InvalidNodeID();
+    // @notice signature: 0x87654321
+    error ExceedsMaxNodeIDs();
+    // @notice signature: 0x13579246
+    error DuplicateNodeID();
```
Comment (zzzckck) [inline contracts/StakeHub.sol:94]: it is the keccak256 hash of the error's signature.
Keccak-256("InvalidNodeID()") = 6490ffd338eeee962e794cd2a5211ddced9e9fe80c50292a2b361adeac34888b
so `error InvalidNodeID()`'s signature will be `0x6490ffd3`
similarly:
  ExceedsMaxNodeIDs() => `0x246be614`
  DuplicateNodeID() => `0x440bc78e`

File: contracts/StakeHub.sol [hunk 18]
```diff
@@ -1141,4 +1282,19 @@ contract StakeHub is SystemV2, Initializable, Protectable {
 
         return msg.sender;
     }
+
+    function _bep563MsgSender() internal view returns (address) {
+        if (consensusToOperator[msg.sender] != address(0)) {
+            return consensusToOperator[msg.sender];
+        }
+
```
Comment (zzzckck) [inline contracts/StakeHub.sol:1290]: should also consider `agentToOperator`

File: contracts/StakeHub.sol [hunk 19]
```diff
@@ -84,6 +86,12 @@ contract StakeHub is SystemV2, Initializable, Protectable {
     error InvalidAgent();
     // @notice signature: 0x682a6e7c
     error InvalidValidator();
+    // @notice signature: 0x12345678
```
Comment (zzzckck) [inline contracts/StakeHub.sol:89]:
```suggestion
    // @notice signature: 0x6490ffd3
```

File: contracts/StakeHub.sol [hunk 20]
```diff
@@ -84,6 +86,12 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+    error InvalidNodeID();
+    // @notice signature: 0x87654321
```
Comment (zzzckck) [inline contracts/StakeHub.sol:91]:
```suggestion
    // @notice signature: 0x246be614
```

File: contracts/StakeHub.sol [hunk 21]
```diff
@@ -84,6 +86,12 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+    error ExceedsMaxNodeIDs();
+    // @notice signature: 0x13579246
```
Comment (zzzckck) [inline contracts/StakeHub.sol:93]:
```suggestion
    // @notice signature: 0x440bc78e
```

File: contracts/StakeHub.sol [hunk 22]
```diff
@@ -1011,6 +1039,119 @@ contract StakeHub is SystemV2, Initializable, Protectable {
...
+        }
+    }
+
+    /**
+     * @notice Returns all validators with their consensus addresses and registered NodeIDs.
+     * @param validatorsToQuery The operator addresses of the validators.
+     * @return consensusAddresses Array of consensus addresses corresponding to the validators.
+     * @return nodeIDsList Array of NodeIDs for each validator.
+     */
+    function listNodeIDsFor(
```
Comment (zzzckck) [inline contracts/StakeHub.sol:1138]: can we change the method name to `getNodeIDs(...)`, so it would be consistent with `addNodeIDs` and `removeNodeIDs`

**Review by zzzckck — CHANGES_REQUESTED** [review]
will merge this PR once it is fully tested.

File: contracts/StakeHub.sol (hunk 22 above)
Comment (zzzckck) [inline contracts/StakeHub.sol:1138]:
```suggestion
    function getNodeIDs(
```

File: contracts/StakeHub.sol [hunk 23]
```diff
@@ -804,6 +827,11 @@ contract StakeHub is SystemV2, Initializable, Protectable {
             address newStakeHubProtector = value.bytesToAddress(20);
             if (newStakeHubProtector == address(0)) revert InvalidValue(key, value);
             _setProtector(newStakeHubProtector);
+        } else if (key.compareStrings("maxNodeIDs")) {
+            if (value.length != 32) revert InvalidValue(key, value);
+            uint256 newMaxNodeIDs = value.bytesToUint256(32);
+            if (newMaxNodeIDs == 0) revert InvalidValue(key, value);
```
Comment (NathanBSC) [inline contracts/StakeHub.sol:833]: if this restrict exsits
we can delete maxNodeIDsInitialized

**Review by joelythe — APPROVED** [review]

**Review by pythonberg1997 — APPROVED** [review]

**Review by zzzckck — APPROVED** [review]

**Review by galaio — APPROVED** [review]

**Review by buddh0 — APPROVED** [review]
